BH fabric. Take care of 2 risc cores in 1 eth core

daminakaTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/21235

### Problem description
Blackhole eth core has 2 risc cores.

### What's changed
Make fabric_erisc_datamover.cpp kernel to know which risc core is used.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

daminakaTT committed:
BW fabric. Take care of 2 risc cores in 1 eth core

daminakaTT committed:
pass risc_id from generate_edm_kernel, use DataMovementProcessor name

jasondavies commented:
> Created #22237 to track the LLK issue

Is this the correct issue?  I can't see anything relating to LLKs here, wondering if it's a typo?

daminakaTT committed:
Merge branch 'main' into dev/daiki/fabric_edm_bw_support

daminakaTT committed:
Add knob to enable/disable sender/receiver channels (#22183)

### Ticket
https://github.com/tenstorrent/tt-metal/issues/21235

### Problem description
To prepare for Blackhole's eth core enablement, sender and receiver need
to be able to be enabled/disabled conditionally

### What's changed
Add flags for each sender/receiver channels. All are enabled by default
for now

### Checklist
- [ ] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml)
CI passes
- [ ] [Blackhole Post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml)
CI with demo tests passes (if applicable)
- [ ] [Model
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml)
CI passes (if applicable)
- [ ] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml)
CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml)
CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

daminakaTT committed:
use device.arch() instead of is_bh flag

sjameelTT commented:
> > Created #22237 to track the LLK issue
> 
> Is this the correct issue? I can't see anything relating to LLKs here, wondering if it's a typo?

whoops lol, posted it in the wrong issue lol

daminakaTT committed:
prepare more variable for each risc cores

daminakaTT committed:
fix another builder creation path

daminakaTT committed:
fix build error

daminakaTT committed:
configurable context switch interval and inner loop count

daminakaTT committed:
minor fix

daminakaTT committed:
Merge branch 'main' into dev/daiki/fabric_edm_bw_support

SeanNijjar commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1836,64 +1854,69 @@ void kernel_main() {
-                // open connections with available downstream edms
-                downstream_edm_noc_interfaces[edm_index].template open<true, tt::tt_fabric::worker_handshake_noc>();
-                *downstream_edm_noc_interfaces[edm_index].from_remote_buffer_slot_rdptr_ptr = 0;
+            wait_for_notification((uint32_t)edm_status_ptr, tt::tt_fabric::EDMStatus::READY_FOR_TRAFFIC);
```
Some of the changes here (including this one) look wrong but it could be because of the new indentation and github diff being confusing. I'll re-review this after removing the condition mentioned in previous comment.

daminakaTT replied:
Here is just wrapped by big if block.

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/erisc_datamover_builder_helper.hpp:
```diff
@@ -104,11 +104,13 @@ class EdmLineFabricOpInterface {
     static void teardown_edm_fabric(MeshDevice* mesh_device);
 
-    void set_firmware_context_switch_interval(size_t interval);
+    void set_firmware_context_switch_interval(size_t interval, size_t risc_id = std::numeric_limits<size_t>::max());
```
I don't think we need to specify this per risc.

SeanNijjar reviewed (commented):
I think we can remove the complexity caused by adding extra risc support to this PR and for future changes if we move the work-split functionality to a separate component, then the changes should also be a lot simpler and more modular.

daminakaTT commented on tests/tt_metal/tt_fabric/fabric_data_movement/test_basic_1d_fabric.cpp:
```diff
@@ -536,18 +536,22 @@ void RunTestUnicastRaw(BaseFabricFixture* fixture, uint32_t num_hops, RoutingDir
-        .edm_worker_location_info_addr = edm_config.sender_channels_worker_conn_info_base_address[sender_channel],
-        .buffer_size_bytes = edm_config.channel_buffer_size_bytes,
-        .buffer_index_semaphore_id = edm_config.sender_channels_buffer_index_semaphore_address[sender_channel],
+        .edm_buffer_base_addr = edm_config.sender_channels_base_address[risc_id][sender_channel],
```
But this touches FabricEriscDatamoverConfig, so it is inevitable to risc_id (always 0).

SeanNijjar replied:
I think some of my later comments clarify this a little better. We shouldn't need to track these buffer addresses on a _per erisc_ basis (the addresses are to carve up L1, which is shared by both eriscs).

daminakaTT replied:
Later comments you mentioned is to keep these arrays. So we have to access with risc_id even though the address are same for all riscs.

daminakaTT replied:
So all values here need to be per eth core or per risc?
The values can be different when `shared_buffer_mode == false` otherwise same. so In any case this risc_id (== 0)  based access works

daminakaTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1836,64 +1854,69 @@ void kernel_main() {
-            wait_for_notification((uint32_t)edm_local_sync_ptr, num_local_edms - 1);
-            notify_subordinate_routers(
-                edm_channels_mask, local_handshake_master_eth_chan, (uint32_t)edm_local_sync_ptr, num_local_edms);
+    if constexpr (enable_context_switch) {
```
Maybe I intended to do `if (enable_handshake)`, not enable_context_switch
I thought this big block is for handshake

daminakaTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1696,7 +1714,7 @@ void kernel_main() {
                                                             local_sem_for_teardown_from_downstream_edm[edm_index]);
-                if constexpr (is_2d_fabric) {
+                if constexpr (is_2d_fabric && enable_handshake) {
                     // reset the handshake addresses to 0
```
This is for handshake, isn't it?

daminakaTT replied:
oh oK

daminakaTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1356,7 +1372,7 @@ void __attribute__((noinline)) init_local_sender_channel_worker_interfaces(
 
 void kernel_main() {
-    eth_txq_reg_write(DEFAULT_ETH_TXQ, ETH_TXQ_DATA_PACKET_ACCEPT_AHEAD, DEFAULT_NUM_ETH_TXQ_DATA_PACKET_ACCEPT_AHEAD);
+    eth_txq_reg_write(sender_txq_id, ETH_TXQ_DATA_PACKET_ACCEPT_AHEAD, DEFAULT_NUM_ETH_TXQ_DATA_PACKET_ACCEPT_AHEAD);
```
`receiver_txq_id`? `receiver_rxq_id` is the correct one?

daminakaTT committed:
Update tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp

Co-authored-by: Sean Nijjar <sean.nijjar@gmail.com>

daminakaTT committed:
Update tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp

Co-authored-by: Sean Nijjar <sean.nijjar@gmail.com>

daminakaTT committed:
Update tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp

Co-authored-by: Sean Nijjar <sean.nijjar@gmail.com>

daminakaTT committed:
follow comments v1

daminakaTT committed:
teardown once

daminakaTT committed:
Merge branch 'main' into dev/daiki/fabric_edm_bw_support

daminakaTT committed:
fix debug build error

daminakaTT committed:
use fill_array_with_next_n_args

daminakaTT committed:
rollback flags/values for per eth core elements

daminakaTT commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -454,26 +503,26 @@ std::vector<uint32_t> FabricEriscDatamoverBuilder::get_compile_time_args() const
-        config.sender_channels_base_address[3],
-        config.sender_channels_base_address[4],
+        config.sender_channels_base_address[risc_id][0],
+        config.sender_channels_worker_conn_info_base_address[risc_id][0],
```
this has same address across all riscs by default

SeanNijjar replied:
The per risc indexing is not needed.

daminakaTT commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -482,31 +531,44 @@ std::vector<uint32_t> FabricEriscDatamoverBuilder::get_compile_time_args() const
-        config.sender_channels_counters_address[2],
-        config.sender_channels_counters_address[3],
-        config.sender_channels_counters_address[4],
+        config.receiver_channels_counters_address[risc_id][0],
```
this has same address across all riscs by default

nhuang-tt commented on tt_metal/api/tt-metalium/erisc_datamover_builder.hpp:
```diff
@@ -48,9 +48,11 @@ struct FabricEriscDatamoverConfig {
     static constexpr std::size_t max_downstream_edms = std::max(num_downstream_edms, num_downstream_edms_2d);
     static constexpr uint32_t num_virtual_channels = 2;
 
+    static constexpr std::size_t num_riscv_cores = 1;  // 2 for BH
```
This hardware specific value belongs in the HAL. I believe you are trying to create a Fabric router on an Active Ethernet core. You can query it like this
```cpp
tt::tt_metal::MetalContext::instance()
  .hal()
  .get_processor_classes_count(HalProgrammableCoreType::ACTIVE_ETH);
```

nhuang-tt commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -122,98 +127,116 @@ FabricEriscDatamoverConfig::FabricEriscDatamoverConfig(Topology topology) {
-    std::size_t channel_buffer_size_bytes, Topology topology, bool is_dateline) :
+    std::size_t channel_buffer_size_bytes, Topology topology, bool is_dateline, tt::ARCH arch) :
     FabricEriscDatamoverConfig(topology) {
+    this->num_used_riscv_cores = FabricEriscDatamoverConfig::num_riscv_cores;
```
Relevant values to query from the HAL

daminakaTT committed:
use hal API to get the number of risc. change array to vector accordingly

SeanNijjar commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1836,64 +1855,69 @@ void kernel_main() {
-            wait_for_notification((uint32_t)edm_local_sync_ptr, num_local_edms - 1);
-            notify_subordinate_routers(
-                edm_channels_mask, local_handshake_master_eth_chan, (uint32_t)edm_local_sync_ptr, num_local_edms);
+    if constexpr (enable_handshake) {
```
The local chip EDM connection setup process will need tweaking when we enable multiple eriscs but for now, I think this just needs to be modified to close around 1890 after the later call to `notify_subordinate_routers`. (I'll mark it with a separate comment)

SeanNijjar commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -719,10 +784,12 @@ FabricEriscDatamoverBuilder FabricEriscDatamoverBuilder::build(
             }
             // Sender channel 0 uses addresses instead of ids in persistent mode
-            sender_channels_buffer_index_semaphore_id[0] = config.sender_channels_buffer_index_semaphore_address[0];
+            sender_channels_buffer_index_semaphore_id[0] =
```
revert (block)

SeanNijjar commented on ttnn/cpp/ttnn/operations/ccl/erisc_datamover_builder_helper.cpp:
```diff
@@ -124,17 +124,21 @@ EdmLineFabricOpInterface::EdmLineFabricOpInterface(
             edm_builders_forward_direction[src_device->id()].reserve(local_link_cores.size());
             edm_builders_backward_direction[dest_device->id()].reserve(local_link_cores.size());
             for (size_t l = 0; l < this->num_links; l++) {
+                std::vector<tt::tt_fabric::FabricEriscDatamoverBuilder> fwd_edm_builders_for_risc;
```
I think the natural consequence of the previous suggestion is that the changes to this fill will all be revertable (probably, I may have missed a couple lines that aren't). We don't need to add BH support right now, just the knobs. We should have a separate per-risc builder added as next steps (future PRs) and I think it will look different from how it's done here.

daminakaTT commented on tests/ttnn/unit_tests/gtests/ccl/test_fabric_edm_common.hpp:
```diff
@@ -1211,14 +1211,16 @@ void setup_test_with_persistent_fabric(
+                edm_builders_per_conn.begin()->my_noc_x,
+                edm_builders_per_conn.begin()->my_noc_y);
+            for (auto& edm_per_risc : edm_builders_per_conn) {
+                edm_per_risc.connect_to_downstream_edm(edm_per_risc);
```
This is impossible to fix unless `edm_builders_backward_direction` to be another struct inastead of `vector<vector>`

daminakaTT commented on tt_metal/api/tt-metalium/erisc_datamover_builder.hpp:
```diff
@@ -207,7 +226,9 @@ class FabricEriscDatamoverBuilder {
-        bool dateline_connection = false);
+        bool dateline_connection = false,
+        tt::ARCH arch = tt::ARCH::WORMHOLE_B0,
+        size_t risc_id = 0);
```
This is needed for `is_sender_channel_serviced`, `is_receiver_channel_serviced`, `enable_handshake`, `enable_context_switch`, `enable_interrupts` and `iterations_between_ctx_switch_and_teardown_checks`

daminakaTT replied:
This is needed (for now) to extract per risc items from FabricEriscDatamoverConfig

daminakaTT committed:
follow commonts. revert most of risc_id access

daminakaTT committed:
Merge branch 'main' into dev/daiki/fabric_edm_bw_support

yugaoTT commented on tt_metal/impl/dispatch/topology.cpp:
```diff
@@ -1101,18 +1102,22 @@ void build_tt_fabric_program(
-                eth_direction);
-            edm_builders.insert({eth_chan, edm_builder});
+            std::vector<tt::tt_fabric::FabricEriscDatamoverBuilder> edm_builders_for_cores;
+            for (size_t risc_id = 0; risc_id < (device->arch() == ARCH::BLACKHOLE ? 2 : 1); risc_id++) {
```
this can also be pull from HAL?

yugaoTT commented on tt_metal/impl/dispatch/topology.cpp:
```diff
@@ -1182,18 +1187,23 @@ void build_tt_fabric_program(
-                control_plane->routing_direction_to_eth_direction(direction));
-            edm_builders.insert({eth_chan, edm_builder});
+            std::vector<tt::tt_fabric::FabricEriscDatamoverBuilder> edm_builders_for_cores;
+            for (size_t risc_id = 0; risc_id < (device->arch() == ARCH::BLACKHOLE ? 2 : 1); risc_id++) {
```
same above

yugaoTT commented on tt_metal/impl/dispatch/topology.cpp:
```diff
@@ -1220,18 +1230,20 @@ void build_tt_fabric_program(
                 auto& edm_builder2 = edm_builders.at(eth_chan_dir2);
-                edm_builder1.connect_to_downstream_edm(edm_builder2);
-                edm_builder2.connect_to_downstream_edm(edm_builder1);
+                for (size_t risc_id = 0; risc_id < (device->arch() == ARCH::BLACKHOLE ? 2 : 1); risc_id++) {
```
same above

daminakaTT committed:
use hal to take risc core count

daminakaTT committed:
cleanup

daminakaTT committed:
add remark for follow up PR

daminakaTT commented:
Per risc update
https://github.com/tenstorrent/tt-metal/pull/22542

daminakaTT committed:
Having per risc fabric config (#22542)

### Ticket
https://github.com/tenstorrent/tt-metal/issues/21235

### Problem description
FabricEriscDatamoverConfig have per risc parameters by vector.

### What's changed
Replace the vectors by vector of FabricRiscConfig which hold all the
params

### Checklist
- [ ] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml)
CI passes
- [ ] [Blackhole Post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml)
CI with demo tests passes (if applicable)
- [ ] [Model
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml)
CI passes (if applicable)
- [ ] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml)
CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
CI passes (if applicable) See [recommended dev
flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

daminakaTT committed:
cleanup and rollback accidentally removed line

SeanNijjar commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -49,6 +49,16 @@ namespace tt::tt_fabric {
+    enable_interrupts(true),
+    iterations_between_ctx_switch_and_teardown_checks(
+        FabricEriscDatamoverConfig::default_iterations_between_ctx_switch_and_teardown_checks) {
+    this->is_sender_channel_serviced.fill(true);
```
maybe just [0] should be true? Otherwise by default, I think it would break BH because both riscs would try servicing the same channels.

daminakaTT committed:
remove useless variables

SeanNijjar commented on tt_metal/fabric/hw/inc/edm_fabric/1d_fabric_constants.hpp:
```diff
@@ -170,12 +170,23 @@ constexpr size_t sender_3_completed_packet_header_cb_size_headers =
-
-constexpr size_t SPECIAL_MARKER_0_IDX = MAIN_CT_ARGS_START_IDX + 59;
+constexpr std::array<bool, MAX_NUM_SENDER_CHANNELS> is_sender_channel_serviced =
+    fill_array_with_next_n_args<bool, MAIN_CT_ARGS_START_IDX + 56, MAX_NUM_SENDER_CHANNELS>();
```
because `MAX_NUM_SENDER_CHANNELS` is configurable, then I think it is preferred to do something like
```suggestion
constexpr size_t sender_channel_serviced_args_idx = MAX_NUM_SENDER_CHANNELS = 56;
constexpr std::array<bool, MAX_NUM_SENDER_CHANNELS> is_sender_channel_serviced =
    fill_array_with_next_n_args<bool, MAIN_CT_ARGS_START_IDX + 56, MAX_NUM_SENDER_CHANNELS>();

constexpr size_t receiver_channel_serviced_args_idx = sender_channel_serviced_args_idx + is_sender_channel_serviced.size();
constexpr std::array<bool, MAX_NUM_RECEIVER_CHANNELS> is_receiver_channel_serviced =
    fill_array_with_next_n_args<bool, receiver_channel_serviced_args_idx, MAX_NUM_RECEIVER_CHANNELS>();

constexpr size_t NEXT_ARGS_BLOCK_IDX = receiver_channel_serviced_args_idx + is_receiver_channel_serviced.size();
```
(or if there's a prettier way, then go with that instead)

Then the following args are indexing with:
```
... = get_compile_time_arg_val(NEXT_ARGS_BLOCK_IDX);
... = get_compile_time_arg_val(NEXT_ARGS_BLOCK_IDX + 1);
... = get_compile_time_arg_val(NEXT_ARGS_BLOCK_IDX + 2);
...
```

daminakaTT committed:
Merge branch 'main' into dev/daiki/fabric_edm_bw_support

SeanNijjar reviewed (approved):
conditionally approving. I think there is one bug to fix and then how the CT args are indexed I think should be updated too. There were a couple ctors that took risc_id still that don't need it but we discussed that sideband.

Please run the appropriate fabric functional and performance tests.

daminakaTT committed:
more cleanup

daminakaTT committed:
follow comments

daminakaTT commented:
fabric test passes
https://github.com/tenstorrent/tt-metal/actions/runs/15284478711

aagarwalTT commented on tt_metal/api/tt-metalium/erisc_datamover_builder.hpp:
```diff
@@ -267,6 +285,9 @@ class FabricEriscDatamoverBuilder {
 
     std::array<size_t, FabricEriscDatamoverConfig::num_sender_channels> local_sender_channels_connection_info_addr;
 
+    std::array<size_t, FabricEriscDatamoverConfig::num_sender_channels> is_sender_channel_serviced;
+    std::array<size_t, FabricEriscDatamoverConfig::num_receiver_channels> is_receiver_channel_serviced;
```
why do we need them here again since we already have these arrays in `FabricRiscConfig`?

daminakaTT replied:
My VSCode is dead.... search did not catch these....

daminakaTT committed:
follow comments

daminakaTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1840,34 +1863,36 @@ void kernel_main() {
-        *edm_status_ptr = tt::tt_fabric::EDMStatus::LOCAL_HANDSHAKE_COMPLETE;
+        *edm_status_ptr = tt::tt_fabric::EDMStatus::REMOTE_HANDSHAKE_COMPLETE;
+
+        if constexpr (wait_for_host_signal) {
```
I think signal is sent only once from host. I have not touched signaling code on host

aagarwalTT commented on tt_metal/fabric/impl/kernels/edm_fabric/fabric_erisc_datamover.cpp:
```diff
@@ -1840,34 +1863,36 @@ void kernel_main() {
         }
 
-        *edm_status_ptr = tt::tt_fabric::EDMStatus::LOCAL_HANDSHAKE_COMPLETE;
+        *edm_status_ptr = tt::tt_fabric::EDMStatus::REMOTE_HANDSHAKE_COMPLETE;
```
How are we addressing the synchronization b/w the risc cores on the same ethernet 'core'? Should we have a master-subordinate construct for the risc cores as well? Currently, the host only signals/polls the risc which has `enable_ethernet_handshake` set to true, but how would we know the status of other risc cores?

SeanNijjar replied:
When we add BH support this will need to be added. There's nowhere to "put" that new functionality right now but yea that will be needed as (soon) future work

daminakaTT commented:
all post-commit tests
https://github.com/tenstorrent/tt-metal/actions/runs/15286575689
Blackhole
https://github.com/tenstorrent/tt-metal/actions/runs/15286574551
T3K
https://github.com/tenstorrent/tt-metal/actions/runs/15286577926

omilyutin-tt commented on tt_metal/api/tt-metalium/erisc_datamover_builder.hpp:
```diff
@@ -127,6 +131,9 @@ struct FabricEriscDatamoverConfig {
     std::size_t num_used_sender_channels = 0;
     std::size_t num_used_receiver_channels = 0;
     std::size_t num_fwd_paths = 0;
+    std::size_t sender_txq_id;
+    std::size_t receiver_txq_id;
```
Default initialize these to 0

omilyutin-tt commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -49,6 +49,26 @@ namespace tt::tt_fabric {
+        this->enable_context_switch = false;
+        this->enable_interrupts = false;
+    } else {
+        TT_ASSERT(false);
```
`TT_THROW` 

or better use an exhaustive switch

daminakaTT commented on tt_metal/api/tt-metalium/erisc_datamover_builder.hpp:
```diff
@@ -147,6 +154,17 @@ struct FabricEriscDatamoverConfig {
     FabricEriscDatamoverConfig(Topology topology = Topology::Linear);
 };
 
+class FabricRiscConfig {
+public:
+    FabricRiscConfig(uint32_t risc_id);
+    bool enable_handshake;
+    bool enable_context_switch;
+    bool enable_interrupts;
+    size_t iterations_between_ctx_switch_and_teardown_checks;
+    std::array<bool, FabricEriscDatamoverConfig::num_sender_channels> is_sender_channel_serviced;
+    std::array<bool, FabricEriscDatamoverConfig::num_receiver_channels> is_receiver_channel_serviced;
+};
```
Why? all member in FabricEriscDatamoverConfig are public

daminakaTT replied:
if public is OK, is just use struct. using class is just my habit. My question is why private is needed even though FabricEriscDatamoverConfig is public

blozano-tt commented on tt_metal/api/tt-metalium/erisc_datamover_builder.hpp:
We do now!
https://github.com/tenstorrent/tt-metal/issues/22687

daminakaTT committed:
follow comments
